### PR TITLE
Initial support for NPA and AFS archives

### PR DESCRIPTION
**Pull request by redoste**
Everything seems to work on both Windows and Linux, all NPA and AFS files can be copied as is from the disc to the content root.
Only `nss.npa` needs to be extracted because the bytecode compiler doesn't use the `ContentManager`.

The way the mounts are declared in `Game.json` is quite ugly but I had no other idea that allowed enough flexibility for the edge cases (two archives mounted on `dx`, `dx` archives are called `movie` in `system.npa` and the recursion of the `voice` archives).

### CODE REVIEW

**Comment by sa-exe** [comment]
Okay, that's seriously impressive. Thank you!
I'll try not to nitpick too much.

File: src/NitroSharp/Content/AFSFile.cs [hunk 1]
```diff
@@ -0,0 +1,170 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.MemoryMappedFiles;
```
Comment (sa-exe) [inline src/NitroSharp/Content/AFSFile.cs:4]: What's your rationale for using memory-mapped I/O?

File: src/NitroSharp/Content/AFSFile.cs [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        {
+            if (!_disposed)
+            {
+                if (disposing && _archiveOffset == 0)
+                {
+                    _mmFile?.Dispose();
+                }
+                _disposed = true;
+            }
+        }
```
Comment (sa-exe) [inline src/NitroSharp/Content/AFSFile.cs]: It seems you were trying to follow the original guidelines for implementing `Dispose` which were written way back in the day and also happen to be incredibly confusing.
As long as the class doesn't _directly_ own unmanaged resources (OS handles, native memory etc), there's never a need for implementing a finalizer.
`_mmFile` in this case is a _managed_ resource ⁠— it's an instance of a regular managed type, not a native pointer or a handle. You don't need to overthink it: if something is a .NET type implementing `IDisposable`, then it's always considered a managed resource, even if it uses unmanaged resources under the hood or implements a finalizer.

You don't really need two `Dispose` methods either, as long as you stick to a simple rule throughout the codebase ⁠— avoid mixing both managed and unmanaged resources in the same type. You can also go one step further and mark the type as `sealed`. In fact, that's what I do for all internal types unless there's an actual need to do otherwise.

Did I mention there's also no need for a `_disposed` flag? At least in the case of the standard library types, calling `Dispose` twice should never explode in your face (as in, `CloseHandle`, `free` and similar native functions won't actually be called twice). If you follow the same rule in your code, there's hardly ever a need for a boolean flag.

 To summarize, this code can be safely replaced with the following:
```csharp
public void Dispose()
{
    _mmFile.Dispose();
}
```

Check out [Stephen Cleary's blog](https://blog.stephencleary.com/2009/08/how-to-implement-idisposable-and.html) if you want to learn more about this topic.

File: src/NitroSharp/Content/AFSFile.cs [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+                    uint offset = reader.ReadUInt32();
+                    uint size = reader.ReadUInt32();
+                    _files[fileIndex] = (offset, size);
+                }
+
+                uint stringOffset = reader.ReadUInt32();
+                stream.Seek(stringOffset, SeekOrigin.Begin);
+                for (uint fileIndex = 0; fileIndex < _entriesCount; fileIndex++)
+                {
+                    byte[] name = new byte[32];
```
Comment (sa-exe) [inline src/NitroSharp/Content/AFSFile.cs]: Some people may call this "premature optimization", but if you can easily avoid allocating heap objects in a loop without resorting to hacky or less readable code, then it's really a no-brainer to me. It should be safe to move the allocation out of the loop. In fact, if it's only 32 bytes, is there even a need for a heap allocation? Here's how you can avoid one:
```csharp
Span<byte> name = stackalloc byte[32];
```

File: src/NitroSharp/Content/AFSFile.cs [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+                    reader.Read(name, 0, 32);
+                    string decodedName = _encoding.GetString(name).ToLowerInvariant();
+                    _builtinFileNames[decodedName] = fileIndex;
+
+                    byte[] unk = new byte[16];
```
Comment (sa-exe) [inline src/NitroSharp/Content/AFSFile.cs]: Same thing

File: src/NitroSharp/Content/AFSFile.cs [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+            {
+                if (disposing && _archiveOffset == 0)
+                {
+                    _mmFile?.Dispose();
+                }
+                _disposed = true;
+            }
+        }
+
+        public void OpenArchive()
```
Comment (sa-exe) [inline src/NitroSharp/Content/AFSFile.cs]: I don't see a reason to delay proper initialization — it only introduces unnecessary illegal state ("archive not opened"). If the logic seems to "heavy" to be placed inside a constructor, it's common to make a static method called Load or Open that simply calls a _private_ constructor that actually does all the initialization.

File: src/NitroSharp/Content/ArchiveManager.cs [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+        public ArchiveException(string format, string message)
+            : base($"{format} : {message}")
+        {
+        }
+
+    }
+
+    internal class ArchiveManager : IDisposable
+    {
+        public static readonly string DefaultEncoding = "shift-jis";
```
Comment (sa-exe) [inline src/NitroSharp/Content/ArchiveManager.cs]: This should either be a `const` or a magic string.

File: src/NitroSharp/Content/ArchiveManager.cs [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+
+        private VFSNode _root;
+        private Encoding _encoding;
```
Comment (sa-exe) [inline src/NitroSharp/Content/ArchiveManager.cs]: Make them `readonly`.

File: src/NitroSharp/Content/ArchiveManager.cs [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+                    {
+                        archive = constructor(mmFile, _encoding);
+                        archive.OpenArchive();
+                        break;
+                    }
+                    catch
+                    {
+                        continue;
+                    }
+                }
```
Comment (sa-exe) [inline src/NitroSharp/Content/ArchiveManager.cs]: This seems overly generic to me. We know there's only gonna be 2 archive formats, so is there really a need for this much indirection? Not a fan of using exceptions for this either — you're actually expecting the load operation to fail, there's nothing "exceptional" about it. Consider introducing a static method called `TryLoad` instead.

...Hold on, why can't we just look at the file extension?

File: src/NitroSharp/Content/ArchiveManager.cs [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+        {
+            if (encoding == null)
+            {
+                try
+                {
+                    encoding = Encoding.GetEncoding(DefaultEncoding);
+                }
+                catch (ArgumentException)
+                {
+                    Encoding.RegisterProvider(CodePagesEncodingProvider.Instance);
```
Comment (sa-exe) [inline src/NitroSharp/Content/ArchiveManager.cs]: Consider calling `Encoding.RegisterProvider` in the static constructor instead of catching `ArgumentException`. See [SourceText.cs](https://github.com/CommitteeOfZero/nitrosharp/blob/b5229145a0129624103c6d19fd639623be73c056/src/NitroSharp.NsScript/SourceText.cs#L13).

File: src/NitroSharp/Content/ArchiveManager.cs [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+            {
+                if (disposing)
+                {
+                    _root?.Dispose();
+                }
+                _disposed = true;
+            }
+        }
+
+        public void AddMounts(MountPoint[] mountPoints, string rootDirectory, ContentManager content)
```
Comment (sa-exe) [inline src/NitroSharp/Content/ArchiveManager.cs]: Again, I fail to see a reason to delay initialization. This is something that can be done in the constructor.

File: src/NitroSharp/Content/ArchiveManager.cs [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+                archive.OpenArchive();
+                if (mountPoint.FileNamesIni != null)
+                {
+                    archive.LoadFileNames(content.OpenStream(mountPoint.FileNamesIni));
+                }
+                head.MountedArchives.Add(archive);
+            }
+        }
+
+        private (IArchiveFile? parentArchive, string? archivePath) LocateFileInArchives(string path)
```
Comment (sa-exe) [inline src/NitroSharp/Content/ArchiveManager.cs]: ``(IArchiveFile parentArchive, string archivePath)?`` seems like a better return type since they're either both null or not null.

File: src/NitroSharp/Content/ArchiveManager.cs [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+    {
+        public void OpenArchive();
+        public Stream OpenStream(string path);
+        public bool Contains(string path);
+    }
+
+    internal class VFSNode : IDisposable
+    {
+        public List<IArchiveFile>? MountedArchives { get; set; }
+        public Dictionary<string, VFSNode>? Children { get; set; }
```
Comment (sa-exe) [inline src/NitroSharp/Content/ArchiveManager.cs]: Unfortunately, I'm not yet familiar with NPA and AFS, so I can't really tell if it needs to be this complicated.

File: src/NitroSharp/Content/ArchiveManager.cs [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+                    {
+                        foreach (IArchiveFile mountedArchive in MountedArchives)
+                        {
+                            mountedArchive.Dispose();
+                        }
+                    }
+                }
+                _disposed = true;
+            }
+        }
```
Comment (sa-exe) [inline src/NitroSharp/Content/ArchiveManager.cs]: There's no need for a `Dispose` method of any kind — a node doesn't _own_ the archives it points to.

**Review by sa-exe — CHANGES_REQUESTED** [review]

File: src/NitroSharp/Content/AFSFile.cs (hunk 1 above)
Comment (redoste) [reply src/NitroSharp/Content/AFSFile.cs:4]: It's easy to create a `Stream` restricted to a part of an archive using `CreateViewStream` and thus having a stream that *"looks like"* the one of a normal file, it also keeps all the archives opened and let the OS manage what part of the file should be kept in RAM and what could be freed without too much overhead, so this seemed the easiest and cleanest way to do it.

File: src/NitroSharp/Content/AFSFile.cs (hunk 2 above)
Comment (redoste) [reply src/NitroSharp/Content/AFSFile.cs]: I'm not that familiar with C# and .NET, so I just followed the recommendations in Microsoft documentation (I'm not that surprised that its outdated). Thanks for the advice.

File: src/NitroSharp/Content/AFSFile.cs (hunk 5 above)
Comment (redoste) [reply src/NitroSharp/Content/AFSFile.cs]: I've fixed the issue but I had to leave an `if (_files == null)` in `AFSFile.GetFile` overwise the compiler will continue to issue warnings. Maybe we should not use an ArchiveException in this case. Is there any exception that looks like "this is absolutly not normal that we got here" ?

File: src/NitroSharp/Content/ArchiveManager.cs (hunk 8 above)
Comment (redoste) [reply src/NitroSharp/Content/ArchiveManager.cs]: This is defenitly personal preference but I just prefer to trust magic bytes instead of the extension. For example, the subarchives of `voice` do not have `.afs` in their names. A cleaner way might be to map magic bytes to the classes instead of just using a list but since there is only 2 formats right now, I think it's the cleanest way to do it.

I also chose this structure because it's the most flexible, while noah only uses AFS and NPA I have no idea if other N2 games support other formats. I know the goal right now is just Xbox 360 Noah, but just this small future-proofing doesn't seem overkill.

File: src/NitroSharp/Content/ArchiveManager.cs (hunk 9 above)
Comment (redoste) [reply src/NitroSharp/Content/ArchiveManager.cs]: I did it that way because it might have already been called by the bytecode compiler if it ran, but you're right, running it twice in that situation isn't a problem and makes the code clearer.

File: src/NitroSharp/Content/ArchiveManager.cs (hunk 10 above)
Comment (redoste) [reply src/NitroSharp/Content/ArchiveManager.cs]: Here the delayed initialization is required because `ArchiveManager` uses `ContentManager` to read the INI files in `system.npa`. We can bypass this by assuming that those files will always be contained in an archive but I think it's better to support hybrid situations where both archives and extracted files are used.

File: src/NitroSharp/Content/ArchiveManager.cs (hunk 12 above)
Comment (redoste) [reply src/NitroSharp/Content/ArchiveManager.cs]: This has nothing to do with the archive formats themselves, it's the cleanest way I had found to accomodate the weird edge cases (like 2 archives mounted at the same place). This structure is quite expensive to build because it uses a lot of allocation but it's only done once and should be pretty fast to use during the game. If you've got any better idea for this, I will defenitly implement it, I'm bad at data structures.

File: src/NitroSharp/Content/ArchiveManager.cs (hunk 13 above)
Comment (redoste) [reply src/NitroSharp/Content/ArchiveManager.cs]: That's not how I see it, no one else `Dispose` those since `ArchiveManager` only `Dispose` the root node. I'm probably missing something with my poor C# knowledge.

**Comment by redoste** [comment]
As I said I'm not familiar with .NET and, to be honest, this is my first *"real"* thing with it (okay, I admit it might be dumb to do that in a PR), so don't be afraid to nitpick too much, I had no real idea of how memory management worked in C# before you talked about `Dispose` and didn't even knew that `static` constructors were a thing.

File: src/NitroSharp/Content/AFSFile.cs (hunk 5 above)
Comment (sa-exe) [reply src/NitroSharp/Content/AFSFile.cs]: It's still theoretically possible for ``AFSFile`` to be in an invalid state since ``_files`` is a) nullable b) mutable. As far as the compiler is concerned, you haven't fixed a thing 🙂 
There's nothing preventing you from initializing it in the constructor like this:
```csharp
private AFSFile(
    MemoryMappedFile mmFile,
    uint offset,
    Encoding encoding,
    (uint offset, uint size)[] files,
    Dictionary<string, uint> builtinFileNames)
{
    _iniFileNames = new Dictionary<string, uint>();
    _mmFile = mmFile;
    _archiveOffset = offset;
    _encoding = encoding;
    _files = files;
    _builtinFileNames = builtinFileNames;
}
```

Notice that this constructor takes ``offset`` as a parameter. There's really no need to have 2 almost identical constructors if you can call this one with ``(parent._mmFile, parent._archiveOffset + parent.GetFile(name).offset, encoding, files, builtinFileNames)``

As for the ``OpenArchive`` method, it's longer necessary for it to exist. Conceptually, it's something that happens as part of *loading* the archive, right? You can have one TryLoad method that takes care of everything:
```csharp
public static AFSArchive? TryLoad(MemoryMappedFile mmFile, Encoding encoding, uint archiveOffset = 0)
{
    // contents of OpenArchive
    if (!magic.SequenceEqual(Magic))
    {
        return null;
    }
    return new AFSArchive(mmFile, archiveOffset, encoding, files, builtinFileNames);
}
```

As well as a Load method that throws an exception:
```csharp
public static AFSFile Load(MemoryMappedFile mmFile, Encoding encoding, uint archiveOffset = 0)
{
    return TryLoad(mmFile, encoding, archiveOffset) is AFSFile file
        ? file
        : throw new ArchiveException("AFS", "Unknown magic");
}
```

File: src/NitroSharp/Content/ArchiveManager.cs (hunk 13 above)
Comment (sa-exe) [reply src/NitroSharp/Content/ArchiveManager.cs]: No, you're not missing anything. I was wrong 🙂

File: src/NitroSharp/Content/ArchiveManager.cs (hunk 8 above)
Comment (sa-exe) [reply src/NitroSharp/Content/ArchiveManager.cs]: If you're gonna go down this route, you might as well go all the way 🙂 
```csharp
IArchiveFile? archive = loadMethods
    .Select(load => load(mmFile, _encoding))
    .FirstOrDefault(x => x is not null);
```

**Comment by sa-exe** [comment]
What do you mean "dumb"? I really meant it when I said this PR was impressive. If not for the finalizers in your code, I would probably just merge it and take care of all the stylistic things myself. That is still an option if you're already tired of reading my comments or simply don't care enough about .NET and C#.

File: src/NitroSharp/Content/AFSFile.cs (hunk 5 above)
Comment (redoste) [reply src/NitroSharp/Content/AFSFile.cs]: I see, it's sometimes hard for me to wrap my head around the OOP mindset.

I modified `NPAFile` in the same way to keep a symmetry between the formats.
In `NPAFile` we can throw an exception for multiple reasons, I've setup `TryLoad` to only return `null` when the magic is invalid because that should be the only *"normal"* case, the others represent malformed or unsupported files.

**Comment by redoste** [comment]
No I care about it, what I meant by that was that it might not be well received by some people to do your first tries with an unknown technology in a PR to a public repo.

> I really meant it when I said this PR was impressive.

Thank you :meow:

File: src/NitroSharp/Content/AFSFile.cs (hunk 5 above)
Comment (sa-exe) [reply src/NitroSharp/Content/AFSFile.cs]: I basically asked you to make illegal states unrepresentable, that's not exactly an OOP thing 🙂

File: src/NitroSharp/Content/ArchiveManager.cs (hunk 10 above)
Comment (sa-exe) [reply src/NitroSharp/Content/ArchiveManager.cs]: Honestly I would probably just merge ``ContentManager`` and ``ArchiveManager``.

File: src/NitroSharp/Content/ArchiveManager.cs [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+                head.MountedArchives = new List<IArchiveFile>();
+            }
+
+            string fullPath = Path.Combine(rootDirectory, mountPoint.ArchiveName);
+            if (File.Exists(fullPath))
+            {
+                Func<MemoryMappedFile,Encoding,IArchiveFile?>[] loadMethods = new Func<MemoryMappedFile,Encoding,IArchiveFile?>[]
+                {
+                    (m, e) => NPAFile.TryLoad(m, e),
+                    (m, e) => AFSFile.TryLoad(m, e),
```
Comment (sa-exe) [inline src/NitroSharp/Content/ArchiveManager.cs]: What exactly was the point of changing ``NPAFile.TryLoad`` to ``(m, e) => NPAFile.TryLoad(m, e)``? I'm not necessarily asking you to revert this change (I can do that myself), just wondering.

File: src/NitroSharp/Content/ArchiveManager.cs (hunk 10 above)
Comment (redoste) [reply src/NitroSharp/Content/ArchiveManager.cs]: I merged them and I wanted to reproduce the same pattern as for the archive formats (i.e. a static `Load` function that calls a private constructor after all the archives are opened) but `OpenStream` needs the object to be completely initialized to work, we would need a static and a non static copy of `OpenStream` and it would be a mess.

I chose to just open the archives in the constructor, I have no idea if calling such a heavy method in the constructor is considered bad practice.

File: src/NitroSharp/Content/ArchiveManager.cs (hunk 14 above)
Comment (redoste) [reply src/NitroSharp/Content/ArchiveManager.cs]: This is purely an aesthetic choice, it looks nice to have the function names aligned.

I thought it would be optimized out by the compiler in Release mode, I looked into the IL and it looks like it isn't (at least on my setup (.NET 5.0.8 on Linux)).

Comment (sa-exe) [reply src/NitroSharp/Content/ArchiveManager.cs]: I also meant ``AFSFile.TryLoad``, why would you change one but not the other? It's no big deal, I'm just confused.

File: src/NitroSharp/Content/ArchiveManager.cs (hunk 10 above)
Comment (sa-exe) [reply src/NitroSharp/Content/ArchiveManager.cs]: It's fine, we're not designing a public API here.

File: src/NitroSharp/Content/ArchiveManager.cs (hunk 14 above)
Comment (redoste) [reply src/NitroSharp/Content/ArchiveManager.cs]: It's because `AFSFile.TryLoad` has `archiveOffset` as an optional argument, you can't directly assign a function with an optional argument into a `Func` so I wrapped it into a lambda.

An other solution might be to declare an other `TryLoad` without the optional argument in `AFSFile` that's just a wrapper to the real `TryLoad`.

Comment (sa-exe) [reply src/NitroSharp/Content/ArchiveManager.cs]: I see, thanks for clarifying.

**Comment by sa-exe** [comment]
It looks good enough for me. I might change a few things here and there, but mostly just to keep the style consistent. Thanks for taking the time to address my feedback 🙂
